### PR TITLE
feat(pkg/envoy): incrementing metric for error codes

### PR DESCRIPTION
**Pull request by jaellio**
**Description**:

Increments the ErrCodeCounter metric for logged 
error codes that are defined by OSM. When an error is logged, the
corresponding error code is added to the log's context and the
ErrCodeCounter is incremented with the error code as a label value.

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [X] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | [ ] |
| Other                      | [ ] |

**Testing done**:

In this PR I added the ErrCodeCounter metric to the list of registered metrics for the osm-injector. To validate that the ErrCodeCounter metric was being exposed at the metrics endpoint, I added an error log line to a method I knew would be executed 5 times. After running the automated demo, I port forwarded the metrics endpoint on the osm-injector pod to verify the ErrcodeCounter metric was present and showed the expected count.
injectorMetricsEndpoint

Then I port forwarded Prometheus to verify the new metric was being scraped and no updates to the Prometheus config map were required.
injectorErrorCodeMetric

I cannot force all of the errors to occur that have been modified in this PR. However, I validated that with the same logging configuration (use of the GetErrCodeWithMetrics method to increment the error code metric and return the error code as a string) the ErrCodeCounter metric was properly exposed and scraped and had the expected count value.

Additionally, I verified that metrics incremented prior to being registered with the Prometheus client are still captured. 

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? No
    -   Did you notify the maintainers and provide attribution?

2. Is this a breaking change? No

### CODE REVIEW

**Review by trstringer — APPROVED** [review]
This looks good. I'm wondering if we could benefit from an e2e test to prove this out. This is a critical component that provides a lot of insight that could be heavily relied upon. I think testing for validity under a certain scenario (if it's easy enough to repro a log entry that increments this metric) would be good to test against any future regressions.

That test is probably not trivial, so I think a follow-up PR would be fine.

**Review by shalier — APPROVED** [review]

**Review by shashankram — COMMENTED** [review]
@jaellio Can we make all the error code metrics changes in 1 PR so the conflicts that arise when others change the log lines and codes are minimal.

**Review by nshankar13 — APPROVED** [review]

File: pkg/certificate/providers/certmanager/certificate_manager.go [hunk 1]
```diff
@@ -163,14 +163,16 @@ func (cm *CertManager) issue(cn certificate.CommonName, validityPeriod time.Dura
...
+		// TODO: Need to push metric?
+		log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrGeneratingPrivateKey)).
 			Msgf("Error generating private key for certificate with CN=%s", cn)
 		return nil, fmt.Errorf("failed to generate private key for certificate with CN=%s: %s", cn, err)
 	}
 
 	privKeyPEM, err := certificate.EncodeKeyDERtoPEM(certPrivKey)
 	if err != nil {
-		log.Error().Err(err).Str(errcode.Kind, errcode.ErrEncodingKeyDERtoPEM.String()).
+		// TODO: Need to push metric?
```
[the remark this answers is not in the document]
Comment (jaellio) [reply pkg/certificate/providers/certmanager/certificate_manager.go:174]: The TODOs note errors that may result in a service restart and therefore the corresponding metrics might not be captured by Prometheus.

[the remark this answers is not in the document]
Comment (shashankram) [reply pkg/certificate/providers/certmanager/certificate_manager.go:174]: The existing TODO comment is quite confusing. I suggest rephrasing this to `TODO(#<gh-issue>): metric might not be scraped before process restart resulting from this error`.

[the remark this answers is not in the document]
Comment (jaellio) [reply pkg/certificate/providers/certmanager/certificate_manager.go:174]: Will fix in follow up PR

**Pull request merged**